File: solutions/scala/robot-name/1/src/main/scala/RobotName.scala
import scala.util.Random
import scala.collection.mutable.Set

class Robot {
  private val letters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
  private val digits = "0123456789"
  private val rnd = Random()
  private val cache: Set[String] = Set()
  var name: String = {
    val n = generateName()
    cache += n
    n
  }
  
  private def choose(chars: String): Char = {
    val index = rnd.nextInt(chars.size)
    chars(index)
  }

  private def generateName(): String = {
    val a = choose(letters)
    val b = choose(letters)
    val c = choose(digits)
    val d = choose(digits)
    val e = choose(digits)
    val n = s"$a$b$c$d$e"
    n
  }

  def reset(): Unit = {
    var proposedName = generateName()
    while (cache.contains(proposedName)) {
      proposedName = generateName()
    }
    this.name = proposedName
    cache += proposedName
  }
}

File: solutions/javascript/robot-name/1/robot-name.js
export class Robot {
  static USED_NAMES = new Set();
  #name;
  #nums = "0123456789";
  #chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

  constructor() {
    this.reset();
  }

  reset() {
    this.#name = this.#generateName();
  }

  get name() {
    return this.#name;
  }

  #generateName() {
    while(true) {
      const a1 = randomChoose(this.#chars);
      const a2 = randomChoose(this.#chars);
      
      const n1 = randomChoose(this.#nums);
      const n2 = randomChoose(this.#nums);
      const n3 = randomChoose(this.#nums);
  
      const generatedName = a1 + a2 + n1 + n2 + n3;
      if (!Robot.USED_NAMES.has(generatedName)) {
        Robot.USED_NAMES.add(generatedName);
        return generatedName;
      }
    }
  }

  static releaseNames() {
    Robot.USED_NAMES.clear();
  }
}

function randomChoose(list) {
  const n = Math.floor(Math.random() * list.length);
  return list[n];
}